numpy/random/_generator.pyx

```diff
@@ -4114,7 +4172,122 @@ cdef class Generator:
 
         return diric
 
-    # Shuffling and permutations:
+    def permuted(self, object x, *, axis=None, out=None):
+        """
+        permuted(x, axis=None, out=None)
+
+        Randomly permute `x` along axis `axis`.
+
+        Unlike `shuffle`, each slice along the given axis is shuffled
+        independently of the others.
+
+        Parameters
+        ----------
+        x : array_like
+            Array to be shuffled.
+        axis : int, optional
+            Slices of `x` in this axis are shuffled. Each slice
+            is shuffled independently of the others.  If `axis` is
+            None, the flattened array is shuffled.
+        out : ndarray, optional
+            If given, this is the destinaton of the shuffled array.
+            If `out` is None, a shuffled copy of the array is returned.
+
+        Returns
+        -------
+        ndarray
+            If `out` is None, a shuffled copy of `x` is returned.
+            Otherwise, the shuffled array is stored in `out`,
+            and `out` is returned
+
+        See Also
+        --------
+        shuffle
+        permutation
+
+        Examples
+        --------
+        Create a `numpy.random.Generator` instance:
+
+        >>> rng = np.random.default_rng()
+
+        Create a test array:
+
+        >>> x = np.arange(24).reshape(3, 8)
+        >>> x
+        array([[ 0,  1,  2,  3,  4,  5,  6,  7],
+               [ 8,  9, 10, 11, 12, 13, 14, 15],
+               [16, 17, 18, 19, 20, 21, 22, 23]])
+
+        Shuffle the rows of `x`:
+
+        >>> y = rng.permuted(x, axis=1)
+        >>> y
+        array([[ 4,  3,  6,  7,  1,  2,  5,  0],  # random
+               [15, 10, 14,  9, 12, 11,  8, 13],
+               [17, 16, 20, 21, 18, 22, 23, 19]])
+
+        `x` has not been modified:
+
+        >>> x
+        array([[ 0,  1,  2,  3,  4,  5,  6,  7],
+               [ 8,  9, 10, 11, 12, 13, 14, 15],
+               [16, 17, 18, 19, 20, 21, 22, 23]])
+
+        To shuffle the rows of `x` in-place, pass `x` as the `out`
+        parameter:
+
+        >>> y = rng.permuted(x, axis=1, out=x)
+        >>> x
+        array([[ 3,  0,  4,  7,  1,  6,  2,  5],  # random
+               [ 8, 14, 13,  9, 12, 11, 15, 10],
+               [17, 18, 16, 22, 19, 23, 20, 21]])
+
+        Note that when the ``out`` parameter is given, the return
+        value is ``out``:
+
+        >>> y is x
+        True
+        """
+
+        cdef int ax
+        cdef np.npy_intp axlen, axstride, itemsize
+        cdef void *buf
+        cdef np.flatiter it
+
+        x = np.asarray(x)
+
+        if out is None:
+            out = x.copy()
+        else:
+            if type(out) is not np.ndarray:
+                raise ValueError('out must be a numpy array')
+            out[...] = x
+
+        if axis is None:
+            self.shuffle(out.flat)
+            return out
```

Comment:
Hmmm, this is probably pretty slow, since it works by using the sequence-like flat-iterator. But I guess that is fine, it would be easy to change.  I am a bit curious whether we are OK with changes in the output if we adapt this to non C-order array. I.e. if we update this at some point, we are OK, if the result differs when `out` is not C-order, right?

Reply:
Should you special case the common C-contiguous non-strided case with a reshape?

Reply:
Does this work at all? I'd worry that `shuffle(out.flat)` makes a copy of `out.flat`

Reply:
@seberg, this is admittedly a "quick and dirty" implementation (where "quick" refers to implementation time, not execution time).  Is there an existing pattern in the NumPy API for treating an n-d array (with arbitrary strides) as a 1-d array that would better here?  I can implement *something*, but it would be great if there is already a nice way to do it.

@bashtage, yes, with `axis=None`, a fast path for the contiguous case (either C- or F-, actually) should be pretty easy.

@eric-wieser, I've tested it using arrays with various strides, and it has worked, but maybe I'm just blind to an obvious counter-example.  Are there specific tests you think would be good to check?

Reply:
@WarrenWeckesser I think the easiest solution is probably to use some nitty-gritty C-API (I assume contiguity does not matter, only 1-D does), there are a few "KeepOrder" cases which this misses, I think `ravel()` may actually support "K", unlike `reshape`, so both have their advantages (and I don't care for which way)

```
to_shuffle = out.reshape(-1, order="A")
if to_shuffle.flags.owndata:
    PyArray_SetWritebackIfCopyBase(to_shuffle, out)

self.shuffle(to_shuffle)

if to_shuffle.flags.writebackifcopy:
    PyArray_ResolveWritebackIfCopy(to_shuffle) 
```
I have not confirmed that this is the correct pattern, it certainly isn't correct code as-is.

Reply:
I am also not sure that checking for "owndata" is a great approach to find out whether the ravel/reshape was in-place, although I doubt it can fail here (since these are all plain `ndarray`s)

Reply:
Thanks @seberg, I wouldn't have known to use those functions.

The `owndata` flag might not be the right way to check whether a copy was made.  In the following, `y` gets a copy of the data, but `y.flags.owndata` is False:

```
In [286]: x = np.arange(24).reshape(3, 8)[:, ::-1]

In [287]: x
Out[287]: 
array([[ 7,  6,  5,  4,  3,  2,  1,  0],
       [15, 14, 13, 12, 11, 10,  9,  8],
       [23, 22, 21, 20, 19, 18, 17, 16]])

In [288]: x.strides
Out[288]: (64, -8)

In [289]: y = x.reshape(-1, order='A')

In [290]: y.flags.owndata
Out[290]: False
```

Reply:
Is`PyArray_SetWritebackIfCopyBase` include in the Cython headers for the NumPy C API? I can't get the code to recognize it as a C function.  If I use `np.PyArray_SetWritebackIfCopyBase`, I get `AttributeError: module 'numpy' has no attribute 'PyArray_SetWritebackIfCopyBase'` when I run the code.  If I add `cimport numpy as cnp` (which I wish we did all the time!) and use `cnp.PyArray_SetWritebackIfCopyBase`, cython report this when building: `_generator.pyx:4277:28: cimported module has no attribute 'PyArray_SetWritebackIfCopyBase'`

Reply:
Answering my own question: `PyArray_SetWritebackIfCopyBase` and `PyArray_ResolveWritebackIfCopy` are not in the NumPy .pxd files provided by cython 0.29.21.

Reply:
Hmmm, would have expected those to exist, but if not, I guess we could either add them or define them just for this, @mattip we probably just forgot to add them to the `pyd` (or cython never add them?)

One thing: I guess you actually have to create a contigous copy (on which you set the writebackifcopy), and then pass the ravelled version (if the reshape/ravel fails).  I am also a bit unsure if the *if copy* logic there

Reply:
In fact, for simplicity you do not need to use `np.PyArray_SetWritebackIfCopyBase` at all: Just use the normal `PyArray_FromArray` with the writeback-if-copy flag and call `PyArray_ResolveWritebackIfCopy` unconditionally.

Reply:
For now, I've added this to `_generator.pyx`:
```
cdef extern from "numpy/arrayobject.h":
    int PyArray_SetWritebackIfCopyBase(np.ndarray, np.ndarray)
    int PyArray_ResolveWritebackIfCopy(np.ndarray)
```

Reply:
I think the intention was to try to keep those as internal APIs as much as possible, so if they are needed putting them in `_generator.pyx` seems like a good solution. In general, shouldn't `nditer` be a better solution?

Reply:
> @eric-wieser, I've tested it using arrays with various strides, and it has worked, but maybe I'm just blind to an obvious counter-example.

I should have looked at the source code - `shuffle` has a special unoptimized code path for shuffling non-array sequences.

Reply:
> In general, shouldn't `nditer` be a better solution?

Agreed - it seems like we could have an nditer loop, and call `_shuffle_raw` inside it.

Reply:
What are the advantages of an `nditer` loop over using `PyArray_IterAllButAxis`?  That's not a rhetorical question--I could use some education here.  Presumably `nditer` (or the corresponding C API) already includes the machinery for handling `axis` being a tuple (is that correct?), but are there other advantages?  Can anyone point to an example of using `nditer` (or the C API) that iterates over slices determined by an `axis` argument?

Reply:
I don't see that works, you need to shuffle the *whole* array, I am not aware that it is possible to shuffle in chunks :).

Reply:
*"I don't see that works,..."*  What does *that* refer to?  The use of `nditer`?

Reply:
Yes, sorry. nditer chunks iteration when it doesn't fit the 1-d slice layout.

Reply:
Sorry, need more coffee apparently, it is possible, since you can request a contiguous array. At that point you are using `NpyIter` for nothing else then the `writebackifcopy` logic, but it may be fair to do that anyway and accept the cost just to avoid `writebackifcopy` API.

Reply:
Wait, are y'all talking about trying to use `nditer` in this special case where `axis=None`?  I assumed that suggestion applied to the main loop below.

Locally (but not pushed yet), I've modified the code to not use the one-liner `self.shuffle(out.flat)` when `axis` is None.  I added special casing for contiguous arrays, so I only use the `writebackifcopy` machinery when `axis` is None, `ndim > 1`, and the array is not contiguous (that is not C-contiguous and not F-contiguous).

Reply:
I pushed an update with new handling of the case `axis=None`.

Reply:
Thanks, for me the current solution is fine, and the discussion can be resolved.